vortex-ipc/src/messages/reader_async.rs

```diff
@@ -0,0 +1,66 @@
+use std::pin::Pin;
+use std::task::{ready, Context, Poll};
+
+use bytes::BytesMut;
+use futures_util::{AsyncRead, Stream};
+use pin_project_lite::pin_project;
+use vortex_error::VortexResult;
+
+use crate::messages::{DecoderMessage, MessageDecoder, NextMessage};
+
+pin_project! {
+    pub struct AsyncMessageReader<R> {
+        #[pin]
+        read: R,
+        buffer: BytesMut,
+        decoder: MessageDecoder,
+        bytes_read: usize,
+    }
+}
+
+impl<R: AsyncRead + Unpin> AsyncMessageReader<R> {
```

Comment:
I don't think you need all the `Unpin` bounds, if you're accessing the reader thru projection?

Reply:
good point, will try to remove them